app/src/main/java/org/mozilla/focus/biometrics/BiometricAuthenticationHandler.kt

```diff
@@ -0,0 +1,159 @@
+    private var cancellationSignal: CancellationSignal? = null
+    private var selfCancelled = false
+    private var keyStore: KeyStore? = null
+    private var keyGenerator: KeyGenerator? = null
+    private var cryptoObject: FingerprintManagerCompat.CryptoObject? = null
+
+    var needsAuth = false; private set
+    var biometricFragment: BiometricAuthenticationDialogFragment? = null; private set
+
+    private val DEFAULT_KEY_NAME = "default_key"
+
+    init {
+        try {
+            keyStore = KeyStore.getInstance("AndroidKeyStore")
+
+        } catch (err: KeyStoreException) {
+            throw RuntimeException("Failed to get instance of KeyStore", err)
+        }
+
+        try {
+            keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
+        } catch (err: NoSuchAlgorithmException) {
+            throw RuntimeException("Failed to get an instance of KeyGenerator", err)
+        } catch (err: NoSuchProviderException) {
+            throw RuntimeException("Failed to get an instance of KeyGenerator", err)
+        }
+
+        createKey(DEFAULT_KEY_NAME, false)
+
+        val defaultCipher: Cipher
+        try {
+            defaultCipher = Cipher.getInstance(KeyProperties.KEY_ALGORITHM_AES + "/"
+                    + KeyProperties.BLOCK_MODE_CBC + "/"
+                    + KeyProperties.ENCRYPTION_PADDING_PKCS7)
+        } catch (err: Exception) {
+            throw RuntimeException(err)
+        }
+
+        if (initCipher(defaultCipher, DEFAULT_KEY_NAME)) {
+            cryptoObject = FingerprintManagerCompat.CryptoObject(defaultCipher)
+        }
+
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private fun initCipher(cipher: Cipher, keyName: String): Boolean {
+        try {
+            keyStore?.load(null)
+            val key = keyStore?.getKey(keyName, null) as SecretKey
+            cipher.init(Cipher.ENCRYPT_MODE, key)
+            return true
+        } catch (err: KeyPermanentlyInvalidatedException) {
+            return false
+        } catch (err: Exception) {
+            throw RuntimeException(err)
+        }
+
+    }
+
+    private fun createKey(keyName: String, invalidatedByBiometricEnrollment: Boolean?) {
+        try {
+            keyStore?.load(null)
+
+            val builder = KeyGenParameterSpec.Builder(keyName,
+                    KeyProperties.PURPOSE_ENCRYPT).setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setUserAuthenticationRequired(true)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                builder.setInvalidatedByBiometricEnrollment(invalidatedByBiometricEnrollment!!)
+            }
+
+            keyGenerator?.init(builder.build())
+            keyGenerator?.generateKey()
+
+        } catch (e: Exception) {
+            throw RuntimeException(e)
```

Comment:
\[detekt\] ERROR: \[exceptions/ThrowRuntimeException\] Caught exception is too generic\. Prefer catching specific Exceptions to handle specific error cases\.